File: 3.SortingAlgorithms/quicksort.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <iterator>
#include <vector>
#include <cstdlib>

using std::function;

namespace sort {

template <class Itr>
Itr median(Itr a, Itr b, Itr c) {
  if (*a < *b)
    if (*b < *c)
      return b;
    else if (*a < *c)
      return c;
    else
      return a;
  else if (*a < *c)
    return a;
  else if (*c < *b)
    return b;
  else
    return c;
}

template <class Itr>
Itr partition(Itr begin, Itr end) {
  // choose pivot
  // Itr ipivot = median(begin, begin + (end - begin) / 2, end - 1);
  // 快速排序理论上只需要用小于号进行比较就可以实现
  // 左边的组全部<=pivot, 右边全部>=pivot, 在接下来的排序中，左边等于pivot的元素会全部移到最右边
  // 右边等于pivot的元素会全部移到最左边，实际上数组仍然是排列好的，但这种排序不能保证稳定，即
  // 原本相等的元素保持其未排序前的相对顺序.
  Itr ipivot = end - 1;
  typename std::iterator_traits<Itr>::value_type pivot = *ipivot;
  Itr l = begin, r = end - 1;
  while (1) {
    while (*l < pivot) ++l;
    while (pivot < *r) --r;
    if(l >= r) return l;
    else{
      std::iter_swap(l, r);
      l++;
      r--;
    }
  }
  return l;
}

template <class Itr>
void quickSort(Itr begin, Itr end) {
  // std::vector<int>& v) {
  // some guard
  if (end <= begin + 1) return;
  Itr mid = partition(begin, end);
  // std::copy(begin, end, std::ostream_iterator<int>(std::cout, " "));
  // std::cout << "\t" << (mid - begin);
  // std::cout << std::endl;
  quickSort(begin, mid);
  quickSort(mid, end);
}

}  // namespace sort

void test_quicksort(int length) {
  std::vector<int> v(length);
  for(auto& i:v) {
    i = std::rand() % length;
  }
  std::cout << length;
  sort::quickSort(v.begin(), v.end());
  std::cout << (std::is_sorted(v.begin(), v.end()) ? "success" : "fail") << std::endl;
}

int main() {
  std::vector<int> nums{70, 100, 200, 2000, 30000, 2};
  for(auto i:nums)
    test_quicksort(i);
  function<void(int, int)> huahua = [&](int l, int r) {
    if (l >= r) return;
    int i = l;
    int j = r;
    int p = nums[l + rand() % (r - l + 1)];
    while (i <= j) {
      while (nums[i] < p) ++i;
      while (nums[j] > p) --j;
      if (i <= j) std::swap(nums[i++], nums[j--]);
    }
    huahua(l, j);
    huahua(i, r);
  };
  // huahua(0, nums.size() - 1);
  // std::copy(nums.begin(), nums.end(), std::ostream_iterator<int>(std::cout, "
  // "));
  // sort::quickSort(nums.begin(), nums.end(), nums);
  return 0;
}

File: 3.SortingAlgorithms/quicksort.py
# quicksort.py
# https://blog.csdn.net/qq_39438086/article/details/95055145
def sort(nums):
    quick_sort(nums, 0, len(nums))

def quick_sort(nums, l, r):
    if r - l <= 1: return
    # print(nums)
    ll, rr = partition(nums, l, r)
    # print("mid: {0}".format(mid))
    # print(ll, rr)
    quick_sort(nums, l, rr+1)
    quick_sort(nums, ll, r)


def partition(nums, l, r):
    r = r - 1
    pivot = nums[r]
    
    while(l <= r):
        while(nums[l] < pivot):
            l = l + 1
        while(pivot < nums[r]):
            r = r - 1
        if (l <= r):
            nums[r], nums[l] = nums[l], nums[r]
            l = l + 1
            r = r - 1
    # 如果仅仅返回l，当l恰好等于函数参数中的r时就会在接下来的quicksort函数中陷入死循环
    # quickSort(nums, l, r)
    # partition(nums, l, r) -> mid == r
    # quickSort(nums, l, mid==r)
    return l, r

def quick_sort2(nums, l, r):
    if r - l <= 1: return
    # print(nums)
    mid = partition2(nums, l, r)
    # print("mid: {0}".format(mid))
    # print(ll, rr)
    quick_sort(nums, l, mid)
    quick_sort(nums, mid, r)

def partition2(nums, l, r):
    """在python中函数返回l和r不是一件难事，但是在c语言或
    c++中这意味着需要为数组开辟存储空间或者声明静态变量

    partition2是一种在底层语言中比较容易实现的方式
    """
    r = r - 1
    pivot = nums[r]

    while True:
        while(nums[l] < pivot):
            l = l + 1
        while(nums[r] > pivot):
            r = r - 1
        if (l >= r):
            return l
        else:
            nums[r], nums[l] = nums[l], nums[r]
            l = l + 1
            r = r - 1
        

def is_sorted(nums):
    ans = True
    for i in range(1, len(nums)):
        ans = ans and (nums[i-1] <= nums[i])
        if not ans: return ans
    return ans

if __name__ == "__main__":
    from random import randint
    lens = [10, 300, 2000, 5023, 10012]
    for length in lens:
        nums = [0] * length
        for i in range(len(nums)):
            nums[i] = randint(0, 60000) % length
        quick_sort2(nums, 0, len(nums))
        print(is_sorted(nums))